### PR TITLE
Small refactor on singular association creation

### PR DESCRIPTION
**Pull request by lazaronixon**
### Motivation / Background

Just a small refactor from my last [pull request](https://github.com/rails/rails/pull/46386). 

### Detail

After the record is saved I was calling [set_new_record](https://github.com/rails/rails/compare/main...lazaronixon:refactor-singular-association-creation?expand=1#diff-17dcb8ddd2493c4699d040dfbedff0fdfecdc409ce1ce6e9008c806e2818997eL60) again (called inside [build](https://github.com/rails/rails/compare/main...lazaronixon:refactor-singular-association-creation?expand=1#diff-17dcb8ddd2493c4699d040dfbedff0fdfecdc409ce1ce6e9008c806e2818997eL24)), in order to just [set the foreign key](https://github.com/rails/rails/blob/061bf3156fb90ac6b8ec255dfa39492cf22d7b13/activerecord/lib/active_record/associations/belongs_to_association.rb#L93). To make the intention clearer here and avoid unnecessary method calls I'm just calling the method that matters `replace_keys`.

### CODE REVIEW

File: activerecord/lib/active_record/associations/singular_association.rb [hunk 1]
```diff
@@ -56,8 +56,8 @@ def set_new_record(record)
         def _create_record(attributes, raise_error = false, &block)
           reflection.klass.transaction do
             record = build(attributes, &block)
-            saved = record.save
-            set_new_record(record)
+            saved  = record.save
```
Comment (casperisfine) [inline activerecord/lib/active_record/associations/singular_association.rb:59]:
```suggestion
            saved = record.save
```

**Comment by lazaronixon** [comment]
Just a question... Why do we replace_keys for `create_association`?

shouldn't it be the same ? 

```ruby
# don't replace keys 
firm = client.build_firm name: "baa"
firm.save

# replace keys
firm = client.create_firm name: "baa"
```

This behavior is ensured by theses tests: 
https://github.com/rails/rails/blob/d4c3f4faea890956cd7ab306e9ee9f95c747c9cb/activerecord/test/cases/associations/belongs_to_associations_test.rb#L1417

https://github.com/rails/rails/blob/d4c3f4faea890956cd7ab306e9ee9f95c747c9cb/activerecord/test/cases/associations/belongs_to_associations_test.rb#L1424

**Comment by lazaronixon** [comment]
Actually If I remove `replace_keys`both tests fail, on the other hand if I add a test to check the same behaviour with `build_firm -> save` it fails.

```ruby
def test_should_set_foreign_key_on_create_association
  client = Client.create! name: "fuu"

  firm = client.build_firm name: "baa"
  firm.save

  assert_equal firm.id, client.client_of
end
```

**Comment by lazaronixon** [comment]
@casperisfine isn't it the correct behavior ? 🤔 
https://github.com/rails/rails/pull/46515/commits/8371a1217a8d35758760d3b68f3e1e4d9c1f5fcd

**Comment by rafaelfranca** [comment]
Can you squash the two commits together?

**Pull request merged**